### PR TITLE
Merging simplification

### PR DESCRIPTION
**Pull request by fberlakovich**
This changeset introduces merge configurations. They are simply combinations of strategies and configure a three way merger instance. As can be seen in the example this drastically eases the use of the merge framework. If you like it this way, I will complement the documentation.

Could you please give me some hints on how to seperate the headers into private and public headers? What would go into kdbmerge.h? The idea would be to make only the configurations public, while the more fine grained strategies stay private.

### CODE REVIEW

**Comment by markus2330** [comment]
start build

**Comment by fberlakovich** [comment]
This pull request will also deal with #177 once finished

File: src/tools/kdb/cmdline.cpp [hunk 1]
```diff
@@ -148,21 +148,14 @@ Cmdline::Cmdline (int argc,
...
-			"                         theirs        .. use always their version in case of conflict\n"
-			"                         base          .. use always the base version in case of conflict\n"
-			"                         newkey        .. merge just those keys added by one side only\n"
-			"                         ourvalue      .. use our value if theirs is unmodified\n"
-			"                         theirvalue    .. use their value if ours is unmodified\n"
+			"                         preserve      .. automerge only those keys where just one side deviates from base (default)\n"
+			"                         ours          .. like preserve, but in case of conflict use our version\n"
+			"                         theirs        .. like preserve, but in case of conflict use their version\n"
+			"                         cut           .. primarily used for import. removes existing keys below the import point and always takes the imported version\n"
+			"                         import        .. primarily used for import. preserves existing keys if they do not existing in the imported keyset. in all other\n"
```
Comment (markus2330) [inline src/tools/kdb/cmdline.cpp]: line too long, spelling: "they do not exist"

File: src/tools/kdb/import.cpp [hunk 2]
```diff
@@ -63,16 +63,9 @@ int ImportCommand::execute(Cmdline const& cl)
 
 	KeySet base;
 
-	// TODO: this should not be neccessary, but
-	// applying threeway strategies to a twoway merge is hard
-	base = KeySet();
+	base = existingKeys;
```
Comment (markus2330) [inline src/tools/kdb/import.cpp]: useless copy: directly use existingKeys (you may rename it) or use KeySet const& base

File: src/tools/kdb/import.cpp [hunk 3]
```diff
@@ -63,16 +63,9 @@ int ImportCommand::execute(Cmdline const& cl)
...
 
-	// TODO: for now we have to position this strategy manually
-	// to avoid meta information loss
-	MetaMergeStrategy metaStrategy(merger);
-	merger.addConflictStrategy(&metaStrategy);
-
-	helper.parseStrategies (cl, merger);
+	helper.configureMerger (cl, merger);
```
Comment (markus2330) [inline src/tools/kdb/import.cpp]: Is this really the same as for merge.cpp?

**Comment by markus2330** [comment]
Please also write docu and release notes for that change.

File: src/tools/kdb/import.cpp (hunk 3 above)
Comment (fberlakovich) [inline src/tools/kdb/import.cpp]: To be honest I didn't read into all the import stuff again when patching this part. However, the tests still pass, so I guess it should be ok. It was the same as merge.cpp even before the merge configuration patch, wasn't it?

Comment (markus2330) [inline src/tools/kdb/import.cpp]: Felix Berlakovich wrote:
> To be honest I didn't read into all the import stuff again when patching
> this part. However, the tests still pass, so I guess it should be ok. It

I am not aware of any tests that check if "kdb import" merges correctly.

> was the same as merge.cpp even before the merge configuration patch,
> wasn't it?

You told me about the differences of two-way merging which made it hard to 
implement "kdb import" ;)

Comment (fberlakovich) [inline src/tools/kdb/import.cpp]: > I am not aware of any tests that check if "kdb import" merges correctly.

check_import.sh should do this

> You told me about the differences of two-way merging which made it hard to implement "kdb import" ;)

Yes, but the differences should lie in the used strategies. For the two way stuff I had to implement a few strange strategies rarely used in any other scenario. But these strange strategies should be bundled in the import specific configurations (i.e. import and cut)

**Comment by markus2330** [comment]
Please rebase, a large change was merged.

**Pull request merged**